File: src/main/java/onion/compiler/toolbox/Classes.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.toolbox;

import java.util.Arrays;
import java.util.Set;
import java.util.TreeSet;

import onion.compiler.IRT;

/**
 * @author Kota Mizushima
 * Date: 2005/07/12
 */
public class Classes {
  private Classes() {
  }

  public static Set<IRT.MethodRef> getInterfaceMethods(IRT.ClassTypeRef type){
    Set<IRT.MethodRef> methods =  new TreeSet(new IRT.MethodRefComparator());
    collectInterfaceMethods(type, methods);
    return methods;
  }
  
  private static void collectInterfaceMethods(IRT.ClassTypeRef type, Set<IRT.MethodRef> set){
    IRT.MethodRef[] methods = type.methods();
    set.addAll(Arrays.asList(methods));
    IRT.ClassTypeRef[] interfaces = type.interfaces();
    for (IRT.ClassTypeRef anInterface : interfaces) {
      collectInterfaceMethods(anInterface, set);
    }
  }
}

File: src/main/scala/onion/compiler/toolbox/Classes.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.toolbox

import java.util.Arrays
import java.util.Set
import java.util.TreeSet
import onion.compiler.TypedAST

/**
 * @author Kota Mizushima
 *
 */
object Classes {
  def getInterfaceMethods(`type`: TypedAST.ClassType): Set[TypedAST.Method] = {
    val methods: Set[TypedAST.Method] = new TreeSet[TypedAST.Method](new TypedAST.MethodComparator)
    collectInterfaceMethods(`type`, methods)
    methods
  }

  private def collectInterfaceMethods(`type`: TypedAST.ClassType, set: Set[TypedAST.Method]): Unit = {
    val methods = `type`.methods
    set.addAll(Arrays.asList(methods:_*))
    val interfaces: Seq[TypedAST.ClassType] = `type`.interfaces
    for (anInterface <- interfaces) {
      collectInterfaceMethods(anInterface, set)
    }
  }
}
